nixos/modules/services/mail/automx2.nix

```diff
@@ -8,6 +8,64 @@
 let
   cfg = config.services.automx2;
   format = pkgs.formats.json { };
+  imapSmtpServerType = lib.types.mkOptionType {
+    name = "imapSmtpServerType";
+    description = "automx2 settings: IMAP/SMTP server configuration";
+    check =
+      x:
+      if !builtins.isAttrs x then
+        false
+      else if !lib.types.str.check x.type then
+        false
+      else if x.type != "imap" && x.type != "smtp" then
+        false
+      else if !lib.types.str.check x.name then
+        false
+      else if !lib.types.port.check x.port then
+        false
+      else
+        true;
```

Comment:
I don't get this

Reply:
Do you refer to the individual checks or the overall approach?
- individual checks: I'll add comments to make them easier to understand
- general approach: my goal was, to properly validate the JSON datastructure that needs to be passed to `/initdb` to catch issues at evaluation instead of at runtime with a failing service startup…
  - Is there a better way than using custom types?
  - Is my approach of using type checks wrong?

Reply:
The general approach I must admit. It would be cool, if we could just validate the config with the upstream program. That would make maintenance a lot easier.